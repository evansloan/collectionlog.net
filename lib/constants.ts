export const DISCORD_URL = 'https://discord.gg/loghunters';
export const GITHUB_URL = 'https://github.com/evansloan/collectionlog.net';
export const GITHUB_URL_API = 'https://github.com/evansloan/collection-log-api';
export const GITHUB_URL_PLUGIN = 'https://github.com/evansloan/collection-log';
export const DOCS_URL_API = 'https://docs.collectionlog.net';
export const PLUGIN_URL = 'https://runelite.net/plugin-hub/show/collection-log';
