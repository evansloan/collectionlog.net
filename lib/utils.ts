import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

/**
 * Combines multiple classes or class arrays into a single string.
 *
 * @param {...ClassValue[]} inputs - The classes or class arrays to be combined.
 * @return {string} - The combined classes as a single string.
 */
export const cn = (...inputs: ClassValue[]): string => {
  return twMerge(clsx(inputs));
};

/**
 * Get the locale of the user's browser.
 *
 * @function
 * @returns {string} The locale of the user's browser.
 */
export const getLocale = (): string => {
  if (navigator.languages != undefined) {
    return navigator.languages[0];
  }
  return navigator.language;
};

/**
 * Formats a date string in the specified locale.
 *
 * @param {string} dateString - The date string to format.
 * @param {string} locale - The locale to use for formatting.
 * @returns {string} The formatted date string in the specified locale.
 */
export const formatDate = (dateString: string, locale: string): string => {
  const date = new Date(Date.parse(dateString));
  const offset = date.getTimezoneOffset();
  const tzDate = new Date(date.getTime() - offset);

  return tzDate.toLocaleString(locale, {
    day: 'numeric',
    month: 'numeric',
    year: 'numeric',
  });
};

/**
 * Converts a number to a string with formatting based on the user's locale.
 *
 * @param {number} n - The number to format.
 * @returns {string} The formatted string representation of the number.
 */
export const formatInt = (n: number): string =>
  new Intl.NumberFormat().format(n);

/**
 * Converts a given string to title case.
 *
 * @param {string} str - The string to be converted to title case.
 * @returns {string} - The string in title case.
 */
export const toTitleCase = (str: string): string => {
  return str.replace(/\w\S*/g, (txt) => {
    return txt.charAt(0).toUpperCase() + txt.slice(1).toLowerCase();
  });
};

/**
 * Replaces the current URL in the browser's history with the specified URL.
 *
 * @param {string} url - The new URL to replace the current one.
 * @returns {void}
 */
export const replaceUrl = (url: string): void => {
  window.history.replaceState(
    { ...window.history.state, as: url, url },
    '',
    url
  );
};

/**
 * Sorts an array of strings alphabetically, ignoring the leading "The" keyword.
 *
 * @param {string[]} vals - The array of strings to be sorted.
 * @returns {string[]} - The sorted array of strings.
 */
export const sortAlphabetical = (vals: string[]): string[] => {
  return vals.sort((a, b) => {
    a = a.replace(/^The /, '');
    b = b.replace(/^The /, '');
    return a.localeCompare(b);
  });
};

export const sortKeys = (unordered: { [k: string]: any }) => {
  return Object.keys(unordered)
    .sort()
    .reduce((obj: { [k: string]: any }, key) => {
      obj[key] = unordered[key];
      return obj;
    }, {});
};

/**
 * Converts a string by replacing spaces with hyphens.
 *
 * @param {string} s - The string to be hyphenated.
 * @returns {string} The hyphenated string.
 */
export const hyphenate = (s: string): string => {
  return s.replace(/ /g, '-');
};
