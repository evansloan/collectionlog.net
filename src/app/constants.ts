export enum AccountType {
  NORMAL = 'NORMAL',
  IRONMAN = 'IRONMAN',
  HARDCORE_IRONMAN = 'HARDCORE_IRONMAN',
  ULTIMATE_IRONMAN = 'ULTIMATE_IRONMAN',
  GROUP_IRONMAN = 'GROUP_IRONMAN',
  HARDCORE_GROUP_IRONMAN = 'HARDCORE_GROUP_IRONMAN'
}

export const discordUrl = 'https://discord.gg/loghunters';
export const gitHubUrl = 'https://github.com/evansloan/collectionlog.net';
export const pluginUrl = 'https://runelite.net/plugin-hub/show/collection-log';

export const expectedMaxUniqueItems = 1487;