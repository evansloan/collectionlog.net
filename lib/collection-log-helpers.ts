import { CollectionLogAPI } from '@/lib/api/collection-log/collection-log-api';
import { sortAlphabetical, toTitleCase } from '@/lib/utils';

const TAB_BOSSES = 'Bosses';
const TAB_RAIDS = 'Raids';
const TAB_CLUES = 'Clues';
const TAB_MINIGAMES = 'Minigames';
const TAB_OTHER = 'Other';
const TABS = [TAB_BOSSES, TAB_RAIDS, TAB_CLUES, TAB_MINIGAMES, TAB_OTHER];

const CUSTOM_PAGE_SORT: { [key: string]: string[] } = {
  [TAB_CLUES]: [
    'Beginner Treasure Trails',
    'Easy Treasure Trails',
    'Medium Treasure Trails',
    'Hard Treasure Trails',
    'Elite Treasure Trails',
    'Master Treasure Trails',
    'Hard Treasure Trails (Rare)',
    'Elite Treasure Trails (Rare)',
    'Master Treasure Trails (Rare)',
    'Shared Treasure Trail Rewards',
  ],
};

/**
 * Sorts the pages in the collection log and updates the collection log tabs accordingly.
 *
 * @param {CollectionLog} collectionLog - The original collection log object.
 * @returns {void}
 */
export const sortCollectionLog = (collectionLog: CollectionLog): void => {
  const sorted: CollectionLogTab = {};
  TABS.forEach((tabName) => {
    sorted[tabName] = {};
    const pageNameSort =
      CUSTOM_PAGE_SORT[tabName] ??
      sortAlphabetical(Object.keys(collectionLog.tabs[tabName]));

    pageNameSort.forEach((pageName) => {
      sorted[tabName][pageName] = collectionLog.tabs[tabName][pageName];
    });
  });

  collectionLog.tabs = sorted;
};

/**
 * Formats the account type or rank type into a human-readable format.
 *
 * @param {AccountType|RankType} accountType - The account type or rank type to be formatted.
 * @returns {string} - The formatted account type or rank type.
 */
export const formatAccountType = (
  accountType: AccountType | RankType
): string => {
  return toTitleCase(accountType.replace(/_/g, ' '));
};

export interface CollectionLogFull {
  collectionLog: CollectionLog;
  settings: UserSettings;
}

/**
 * Retrieves the full collection log for a given username.
 *
 * @param {string} username - The username of the user to retrieve the collection log for.
 * @returns {Promise<CollectionLogFull>} - Object containing the collection log, ranks, and settings.
 */
export const getFullCollectionLog = async (
  username: string
): Promise<CollectionLogFull> => ({
  collectionLog: await CollectionLogAPI.getCollectionLog(username),
  settings: await CollectionLogAPI.getUserSettings(username),
});
