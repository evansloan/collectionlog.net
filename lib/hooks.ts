import { useCallback, useEffect, useState } from 'react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';

import { getLocale } from '@/lib/utils';

interface QueryParams {
  [k: string]: string | undefined | null;
}

export const useQueryParams = () => {
  const router = useRouter();
  const pathname = usePathname();
  const searchParams = useSearchParams();

  return useCallback(
    (updateParams: QueryParams) => {
      const queryParams = new URLSearchParams(searchParams.toString());
      Object.entries(updateParams).forEach(([key, value]) => {
        if (!value) {
          return queryParams.delete(key);
        }
        queryParams.set(key, value);
      });
      router.push(`${pathname}?${queryParams.toString()}`);
    },
    [pathname, router, searchParams]
  );
};

export const useLocale = () => {
  const [locale, setLocale] = useState('en');
  useEffect(() => setLocale(getLocale()), []);

  return locale;
};

interface ViewSection<T> {
  name: string;
  data: T;
}

/**
 * Represents the currently opened collection log tab/page combination
 *
 * @interface
 */
export interface OpenView {
  tab: ViewSection<CollectionLogPage>;
  page: ViewSection<{
    items: CollectionLogItem[];
    killCount: CollectionLogKillCount[] | undefined;
  }>;
}

export const useCollectionLogView = (
  collectionLog: CollectionLog,
  pageName?: string
) => {
  const defaultTabName = 'Bosses';
  const defaultPageName = 'Abyssal Sire';

  const { tabs } = collectionLog;

  /**
   * Retrieves the open view based on the provided page name and collection log tabs.
   * If the page name is not provided, the default open view will be returned.
   *
   * @param {string | undefined} pageName - The name of the page to retrieve the view from.
   * @param {CollectionLogTab} tabs - All available collection log tabs.
   * @return {OpenView} - The open view object.
   */
  const getOpenViewFromPageName = (pageName?: string): OpenView => {
    const defaultTab = tabs[defaultTabName];
    const defaultPage = defaultTab[defaultPageName];

    const defaultOpen = {
      tab: {
        name: defaultTabName,
        data: defaultTab,
      },
      page: {
        name: defaultPageName,
        data: defaultPage,
      },
    };

    if (!pageName) {
      return defaultOpen;
    }

    for (const tabName in tabs) {
      const pages = tabs[tabName];
      const foundName = Object.keys(pages).find(
        (p) => p.toLowerCase() === pageName?.toLowerCase()
      );

      if (foundName) {
        const tab = tabs[tabName];
        const page = tab[foundName];

        return {
          tab: {
            name: tabName,
            data: tab,
          },
          page: {
            name: foundName,
            data: page,
          },
        };
      }
    }

    return defaultOpen;
  };

  const [openView, setOpenView] = useState(getOpenViewFromPageName(pageName));

  const updateViewByTab = (tabName: string) => {
    const tab = tabs[tabName];
    const pageName = Object.keys(tab)[0];
    const page = tab[pageName];

    setOpenView({
      page: {
        name: pageName,
        data: page,
      },
      tab: {
        name: tabName,
        data: tab,
      },
    });
  };

  const updateViewByPage = (pageName: string) => {
    setOpenView(getOpenViewFromPageName(pageName));
  };

  return {
    openView,
    updateViewByTab,
    updateViewByPage,
  };
};
