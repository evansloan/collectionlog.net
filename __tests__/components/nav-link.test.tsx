import renderer from 'react-test-renderer';

import NavLink from '@/components/nav-link';
import { DISCORD_URL } from '@/lib/constants';

import HomeIcon from '@/assets/images/home.png';
import HiscoresIcon from '@/assets/images/stats.png';
import DiscordIcon from '@/assets/images/discord.png';

describe(`${NavLink.displayName} component`, () => {
  const links = [
    {
      name: 'Home',
      icon: HomeIcon,
      href: '/home',
    }, {
      name: 'Hiscores',
      icon: HiscoresIcon,
      href: '/hiscores',
    }, {
      name: 'Join the Log Hunters Discord server',
      icon: DiscordIcon,
      href: DISCORD_URL,
    },
  ];

  test.each(links)(
    'should render with name "$name", icon $icon, href "$href"',
    ({ name, icon, href }) => {
      const render = renderer.create(
        <NavLink icon={icon} href={href}>{name}</NavLink>
      ).toJSON();
      expect(render).toMatchSnapshot();
    });
});