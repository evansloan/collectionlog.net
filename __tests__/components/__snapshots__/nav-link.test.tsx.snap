// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLink component should render with name "Hiscores", icon {"blurDataURL": "data:image/png;base64,imagedata", "height": 40, "src": "/img.jpg", "width": 40}, href "/hiscores" 1`] = `
<a
  className="flex items-center gap-1 p-2 text-lg text-white no-underline hover:bg-accent hover:text-rs-orange hover:transition-colors"
  href="/hiscores"
>
  <img
    alt=""
    data-nimg="1"
    decoding="async"
    height={20}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
    srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
    style={
      {
        "color": "transparent",
      }
    }
    width={20}
  />
  Hiscores
</a>
`;

exports[`NavLink component should render with name "Home", icon {"blurDataURL": "data:image/png;base64,imagedata", "height": 40, "src": "/img.jpg", "width": 40}, href "/home" 1`] = `
<a
  className="flex items-center gap-1 p-2 text-lg text-white no-underline hover:bg-accent hover:text-rs-orange hover:transition-colors"
  href="/home"
>
  <img
    alt=""
    data-nimg="1"
    decoding="async"
    height={20}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
    srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
    style={
      {
        "color": "transparent",
      }
    }
    width={20}
  />
  Home
</a>
`;

exports[`NavLink component should render with name "Join the Log Hunters Discord server", icon {"blurDataURL": "data:image/png;base64,imagedata", "height": 40, "src": "/img.jpg", "width": 40}, href "https://discord.gg/loghunters" 1`] = `
<a
  className="flex items-center gap-1 p-2 text-lg text-white no-underline hover:bg-accent hover:text-rs-orange hover:transition-colors"
  href="https://discord.gg/loghunters"
>
  <img
    alt=""
    data-nimg="1"
    decoding="async"
    height={20}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
    srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
    style={
      {
        "color": "transparent",
      }
    }
    width={20}
  />
  Join the Log Hunters Discord server
</a>
`;
